File: packages/loom/parser/src/lexer.rs
//! Line-oriented scanner for the Loom v3 surface.
//!
//! Walks source bytes line-by-line and emits a [`ScannedLine`] for
//! every non-blank line. Each line is classified by its leading
//! tokens — speaker / choice / divert / declaration opener / heading
//! / property / fence / fallback — so the parser can stitch them
//! into the AST without re-scanning characters.
//!
//! Position-as-syntax (spec §2) is implemented here: indentation is
//! a structural signal, ALL-CAPS lines are speakers, lines starting
//! with `*` / `+` / `->` / `==` / `#` are syntactic.

use crate::diagnostics::{Code, Diagnostic};
use crate::source::{Position, Span};

/// One classified non-blank source line.
#[derive(Clone, Debug)]
pub struct ScannedLine {
    /// 0-based line index in the source file.
    pub line: u32,
    /// Column (in bytes) where the line's content begins, i.e. the
    /// indent. Trailing whitespace is not counted toward the next
    /// line's indent.
    pub indent: u32,
    /// The trimmed content (no leading indent, no trailing newline).
    pub text: String,
    /// Byte offset in source where the line's content begins.
    pub start_byte: u32,
    /// Byte offset in source where the line's content ends.
    pub end_byte: u32,
    pub kind: LineKind,
}

impl ScannedLine {
    pub fn span(&self) -> Span {
        Span::new(
            Position::new(self.line, self.indent, self.start_byte),
            Position::new(
                self.line,
                self.indent + (self.text.len() as u32),
                self.end_byte,
            ),
        )
    }
}

/// Surface classification of a single line. Carries no body parse —
/// the parser splits inline `<…>` / `[…]` / `{…}` content out of
/// `text` when it lowers to the AST.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum LineKind {
    /// `# Title` — first-class header heading. Stores the title text
    /// (everything after `#` and surrounding whitespace).
    Heading(String),
    /// `key: value` — a property line in the header *or* the
    /// contract zone under a `==` opener.
    Property { key: String, value: String },
    /// `== knot_name` (with optional trailing whitespace).
    KnotMarker(String),
    /// `INT.` / `EXT.` Fountain-style scene heading.
    SceneHeading(String),
    /// `let name = expr`.
    LetBinding { name: String, expression: String },
    /// `CHARACTER Name [is Mixin, ...]` and friends.
    DeclarationOpener {
        kind_word: String,
        name: String,
        mixin: Vec<String>,
    },
    /// `* …` or `+ …` choice. `sticky` is `true` for `+`.
    Choice { sticky: bool, text: String },
    /// `-> target [with k: v, …]`, `-> END`, or
    /// `-> (name) ->` (tunnel — phase-2 emits a single `DivertLine`
    /// and the parser disambiguates).
    DivertLine(String),
    /// `<-` tunnel-return marker on its own line.
    TunnelReturn,
    /// ALL CAPS speaker cue — `WREN`, `BELLKEEPER`, `WREN | FISHER`.
    Speaker(String),
    /// `(…)` parenthetical on its own line.
    Parenthetical(String),
    /// Triple-backtick fence opener / closer / inline-pair. The
    /// parser groups multi-line fences.
    Fence { tail: String, inline_close: bool },
    /// `<kind: args>` runtime directive on its own line. Stores the
    /// inner text (without the `<`/`>`). Inline directives embedded in
    /// dialogue / action text are split out by the parser in a later
    /// phase.
    Directive(String),
    /// Fallback: action prose or dialogue continuation. Disambiguated
    /// by the parser based on whether a speaker is currently active.
    Prose(String),
}

/// Scan `source` into a list of non-blank classified lines plus
/// diagnostics. Blank lines (whitespace-only) are dropped here —
/// they're only meaningful for separating paragraphs, which the
/// parser handles via `terminates_paragraph` on the *next* non-blank
/// line.
pub fn scan(source: &str) -> (Vec<ScannedLine>, Vec<Diagnostic>) {
    let (stripped, mut diagnostics) = crate::comments::strip(source);
    let source = stripped.as_str();
    let mut lines = Vec::new();
    let mut byte: u32 = 0;
    for (line_idx, raw) in source.split_inclusive('\n').enumerate() {
        let line_start_byte = byte;
        let line_len = raw.len() as u32;
        byte = byte.saturating_add(line_len);

        let stripped = raw.strip_suffix('\n').unwrap_or(raw);
        let stripped = stripped.strip_suffix('\r').unwrap_or(stripped);

        let (indent, content_offset) = leading_indent(stripped);
        let trimmed = &stripped[content_offset..];
        let trimmed = trimmed.trim_end();
        if trimmed.is_empty() {
            continue;
        }

        if stripped[..content_offset].contains('\t') {
            diagnostics.push(Diagnostic::error(
                Code::L1001TabIndent,
                Span::new(
                    Position::new(line_idx as u32, 0, line_start_byte),
                    Position::new(
                        line_idx as u32,
                        content_offset as u32,
                        line_start_byte + content_offset as u32,
                    ),
                ),
                "indentation uses tabs; Loom v3 indents with spaces only",
            ));
        }

        let start_byte = line_start_byte + content_offset as u32;
        let end_byte = start_byte + trimmed.len() as u32;
        let kind = classify(trimmed, line_idx as u32, start_byte, &mut diagnostics);
        lines.push(ScannedLine {
            line: line_idx as u32,
            indent: indent as u32,
            text: trimmed.to_string(),
            start_byte,
            end_byte,
            kind,
        });
    }
    (lines, diagnostics)
}

/// Count leading-whitespace columns + the byte offset where content
/// begins. A tab counts as one column for classification purposes;
/// the tab-warning is raised separately so the column number still
/// reflects what the writer sees in their editor.
fn leading_indent(line: &str) -> (usize, usize) {
    let mut cols = 0usize;
    let mut byte = 0usize;
    for ch in line.chars() {
        match ch {
            ' ' | '\t' => {
                cols += 1;
                byte += ch.len_utf8();
            }
            _ => break,
        }
    }
    (cols, byte)
}

fn classify(text: &str, line: u32, start_byte: u32, diagnostics: &mut Vec<Diagnostic>) -> LineKind {
    // Header heading: `# Title` (not `##` — that's reserved for
    // future use). We accept `#` followed by whitespace + text.
    if let Some(rest) = text.strip_prefix('#') {
        if !rest.starts_with('#') {
            let title = rest.trim().to_string();
            return LineKind::Heading(title);
        }
    }

    // Knot marker: `== name`.
    if let Some(rest) = text.strip_prefix("==") {
        let name = rest.trim().to_string();
        if name.is_empty() {
            diagnostics.push(Diagnostic::error(
                Code::L1004UnnamedKnot,
                line_span(line, start_byte, text),
                "`==` knot marker is missing a name",
            ));
        }
        return LineKind::KnotMarker(name);
    }

    // Scene heading: `INT.`, `EXT.`, `INT./EXT.`, `I/E`.
    if is_scene_heading(text) {
        return LineKind::SceneHeading(text.to_string());
    }

    // `let name = expr` reactive binding.
    if let Some(rest) = text.strip_prefix("let ") {
        if let Some((name, expr)) = rest.split_once('=') {
            return LineKind::LetBinding {
                name: name.trim().to_string(),
                expression: expr.trim().to_string(),
            };
        }
    }

    // Divert / tunnel-return.
    if let Some(rest) = text.strip_prefix("->") {
        return LineKind::DivertLine(rest.trim().to_string());
    }
    if text == "<-" {
        return LineKind::TunnelReturn;
    }

    // Choice — `*` or `+` followed by required whitespace + text.
    if let Some(rest) = text.strip_prefix('*') {
        if let Some(body) = rest.strip_prefix(' ') {
            return LineKind::Choice {
                sticky: false,
                text: body.trim_start().to_string(),
            };
        }
        if rest.is_empty() {
            diagnostics.push(Diagnostic::error(
                Code::L1003EmptyChoice,
                line_span(line, start_byte, text),
                "`*` choice has no body text",
            ));
            return LineKind::Choice {
                sticky: false,
                text: String::new(),
            };
        }
    }
    if let Some(rest) = text.strip_prefix('+') {
        if let Some(body) = rest.strip_prefix(' ') {
            return LineKind::Choice {
                sticky: true,
                text: body.trim_start().to_string(),
            };
        }
    }

    // Whole-line angle-bracket directive — `<kind: args>` or `<kind>`.
    // Multi-line block-opening directives are handled by the parser via
    // indent.
    if text.starts_with('<') && text.ends_with('>') && text.len() >= 2 && text != "<-" {
        let inner = &text[1..text.len() - 1];
        return LineKind::Directive(inner.to_string());
    }

    // Triple-backtick fence.
    if let Some(rest) = text.strip_prefix("```") {
        let inline_close = rest.ends_with("```") && rest.len() >= 3;
        let tail = if inline_close {
            rest[..rest.len() - 3].to_string()
        } else {
            rest.to_string()
        };
        return LineKind::Fence { tail, inline_close };
    }

    // Declaration opener — `KEYWORD Name [is X, Y]`.
    if let Some(opener) = parse_declaration_opener(text) {
        if opener.name.is_empty() {
            diagnostics.push(Diagnostic::error(
                Code::L1006UnnamedDeclaration,
                line_span(line, start_byte, text),
                format!("`{}` declaration is missing a name", opener.kind_word),
            ));
        }
        return LineKind::DeclarationOpener {
            kind_word: opener.kind_word,
            name: opener.name,
            mixin: opener.mixin,
        };
    }

    // Parenthetical line — the whole line is wrapped in `(…)`.
    if let Some(inner) = strip_parens(text) {
        return LineKind::Parenthetical(inner.to_string());
    }

    // Property line — `key: value` (and `:` precedes everything else).
    if let Some((key, value)) = property_split(text) {
        return LineKind::Property {
            key: key.to_string(),
            value: value.to_string(),
        };
    }

    // Speaker cue — pure ALL CAPS (allowing digits, underscores,
    // spaces, and `|` for cohort-of-speakers `WREN | FISHER`).
    if is_speaker_line(text) {
        return LineKind::Speaker(text.to_string());
    }

    LineKind::Prose(text.to_string())
}

fn line_span(line: u32, start_byte: u32, text: &str) -> Span {
    Span::new(
        Position::new(line, 0, start_byte),
        Position::new(line, text.len() as u32, start_byte + text.len() as u32),
    )
}

fn is_scene_heading(text: &str) -> bool {
    let upper = text.trim_start();
    upper.starts_with("INT.")
        || upper.starts_with("EXT.")
        || upper.starts_with("INT/EXT")
        || upper.starts_with("INT./EXT.")
        || upper.starts_with("I/E ")
}

fn strip_parens(text: &str) -> Option<&str> {
    let inner = text.strip_prefix('(')?.strip_suffix(')')?;
    // Reject `(x)y(z)` patterns where parens aren't balanced as the
    // outermost wrapper. A scan suffices because parens nest only
    // shallowly in performer cues.
    let mut depth = 0i32;
    for (idx, ch) in inner.char_indices() {
        match ch {
            '(' => depth += 1,
            ')' => {
                if depth == 0 && idx + 1 != inner.len() {
                    return None;
                }
                depth -= 1;
            }
            _ => {}
        }
    }
    Some(inner)
}

fn property_split(text: &str) -> Option<(&str, &str)> {
    let colon = text.find(':')?;
    let key = &text[..colon];
    // Property keys are simple identifiers (letters, digits,
    // underscore, hyphen).
    if key.is_empty() || !key.chars().all(is_property_key_char) {
        return None;
    }
    // Require whitespace (or end of line) after the colon — this
    // rejects URL-shaped lines like `https://example.com` from being
    // mis-read as properties.
    let after = &text[colon + 1..];
    if !after.is_empty() && !after.starts_with(char::is_whitespace) {
        return None;
    }
    let value = after.trim();
    Some((key.trim(), value))
}

fn is_property_key_char(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_' || ch == '-'
}

fn is_speaker_line(text: &str) -> bool {
    let mut saw_letter = false;
    for ch in text.chars() {
        match ch {
            'A'..='Z' => saw_letter = true,
            '0'..='9' | '_' | ' ' | '|' => {}
            _ => return false,
        }
    }
    saw_letter
}

struct DeclarationOpener {
    kind_word: String,
    name: String,
    mixin: Vec<String>,
}

fn parse_declaration_opener(text: &str) -> Option<DeclarationOpener> {
    let mut parts = text.splitn(2, char::is_whitespace);
    let kind_word = parts.next()?.to_string();
    crate::ast::DeclarationKind::from_keyword(&kind_word)?;
    let rest = parts.next().unwrap_or("").trim();

    let (name, mixin_clause) = match rest.find(" is ") {
        Some(idx) => (rest[..idx].trim(), Some(rest[idx + 4..].trim())),
        None => (rest, None),
    };

    let mixin = mixin_clause
        .map(|c| c.split(',').map(|s| s.trim().to_string()).collect())
        .unwrap_or_default();

    Some(DeclarationOpener {
        kind_word,
        name: name.to_string(),
        mixin,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn first(text: &str) -> LineKind {
        let (lines, _) = scan(text);
        lines.into_iter().next().unwrap().kind
    }

    #[test]
    fn heading_is_parsed() {
        assert_eq!(first("# Saltmere\n"), LineKind::Heading("Saltmere".into()));
    }

    #[test]
    fn property_requires_identifier_key() {
        match first("entry: opening\n") {
            LineKind::Property { key, value } => {
                assert_eq!(key, "entry");
                assert_eq!(value, "opening");
            }
            other => panic!("expected property, got {other:?}"),
        }
        // URL-shaped lines should fall through to prose.
        assert!(matches!(first("https://example.com\n"), LineKind::Prose(_)));
    }

    #[test]
    fn knot_marker() {
        assert_eq!(
            first("== opening\n"),
            LineKind::KnotMarker("opening".into())
        );
    }

    #[test]
    fn declaration_opener_with_mixin() {
        match first("CHARACTER Wren is Keeper, Combatant\n") {
            LineKind::DeclarationOpener {
                kind_word,
                name,
                mixin,
            } => {
                assert_eq!(kind_word, "CHARACTER");
                assert_eq!(name, "Wren");
                assert_eq!(mixin, vec!["Keeper", "Combatant"]);
            }
            other => panic!("expected declaration opener, got {other:?}"),
        }
    }

    #[test]
    fn speaker_vs_action() {
        assert!(matches!(first("WREN\n"), LineKind::Speaker(_)));
        assert!(matches!(first("Wren turns away.\n"), LineKind::Prose(_)));
    }

    #[test]
    fn choice_kinds() {
        assert!(matches!(
            first("* Ring the bell.\n"),
            LineKind::Choice { sticky: false, .. }
        ));
        assert!(matches!(
            first("+ Keep talking.\n"),
            LineKind::Choice { sticky: true, .. }
        ));
    }

    #[test]
    fn divert_and_tunnel_return() {
        assert_eq!(
            first("-> ringing\n"),
            LineKind::DivertLine("ringing".into())
        );
        assert_eq!(first("<-\n"), LineKind::TunnelReturn);
    }

    #[test]
    fn parenthetical_line() {
        assert_eq!(
            first("(quietly)\n"),
            LineKind::Parenthetical("quietly".into())
        );
    }

    #[test]
    fn let_binding() {
        match first("let trusted = Wren.trusts.Player > 50\n") {
            LineKind::LetBinding { name, expression } => {
                assert_eq!(name, "trusted");
                assert_eq!(expression, "Wren.trusts.Player > 50");
            }
            other => panic!("expected let binding, got {other:?}"),
        }
    }

    #[test]
    fn scene_heading() {
        assert_eq!(
            first("INT. LIGHTHOUSE - DAWN\n"),
            LineKind::SceneHeading("INT. LIGHTHOUSE - DAWN".into())
        );
    }

    #[test]
    fn fence_inline_vs_block() {
        match first("```warn lx14```\n") {
            LineKind::Fence { tail, inline_close } => {
                assert_eq!(tail, "warn lx14");
                assert!(inline_close);
            }
            other => panic!("expected inline fence, got {other:?}"),
        }
        match first("```note\n") {
            LineKind::Fence { tail, inline_close } => {
                assert_eq!(tail, "note");
                assert!(!inline_close);
            }
            other => panic!("expected block fence opener, got {other:?}"),
        }
    }

    #[test]
    fn tab_indent_warns() {
        let (_lines, diags) = scan("\tWREN\n");
        assert!(diags.iter().any(|d| d.code == Code::L1001TabIndent));
    }

    #[test]
    fn line_comment_is_invisible_to_classifier() {
        // The bare `//` line strips to whitespace, becomes blank, gets dropped.
        let (lines, diags) = scan("// rough order: bell, beat\nWREN\n");
        assert!(diags.is_empty());
        assert_eq!(lines.len(), 1);
        assert!(matches!(lines[0].kind, LineKind::Speaker(_)));
    }

    #[test]
    fn trailing_line_comment_is_stripped_from_prose() {
        let (lines, _) = scan("It rang. // pickup pace here\n");
        assert_eq!(lines.len(), 1);
        match &lines[0].kind {
            LineKind::Prose(text) => assert_eq!(text, "It rang."),
            other => panic!("expected prose, got {other:?}"),
        }
    }

    #[test]
    fn block_comment_does_not_eat_following_speaker() {
        let (lines, diags) = scan("/* blocking sketch\nlives across lines */\nWREN\n");
        assert!(diags.is_empty());
        assert_eq!(lines.len(), 1);
        assert!(matches!(lines[0].kind, LineKind::Speaker(_)));
    }

    #[test]
    fn unterminated_block_comment_diagnoses() {
        let (_lines, diags) = scan("/* never closed\nstill open\n");
        assert!(diags
            .iter()
            .any(|d| d.code == Code::L1007UnterminatedBlockComment));
    }

    #[test]
    fn url_in_prose_is_not_mistaken_for_comment() {
        let (lines, _) = scan("See https://example.com/path for details.\n");
        assert_eq!(lines.len(), 1);
        match &lines[0].kind {
            LineKind::Prose(text) => {
                assert!(text.contains("https://example.com/path"));
            }
            other => panic!("expected prose, got {other:?}"),
        }
    }

    #[test]
    fn comment_inside_fence_is_preserved() {
        let (lines, _) = scan("```note\n// stage manager: lights low\n```\n");
        // Three fence lines (opener, content, closer) survive.
        assert_eq!(lines.len(), 3);
        // Middle line is a prose line carrying the literal `//`.
        match &lines[1].kind {
            LineKind::Prose(text) => assert!(text.starts_with("//")),
            other => panic!("expected prose inside fence, got {other:?}"),
        }
    }

    #[test]
    fn blank_lines_are_dropped() {
        let (lines, _) = scan("# T\n\n\nentry: x\n");
        assert_eq!(lines.len(), 2);
        assert!(matches!(lines[0].kind, LineKind::Heading(_)));
        assert!(matches!(lines[1].kind, LineKind::Property { .. }));
    }
}

File: packages/loom/core/src/parser/lexer.ts
//! Line-oriented scanner for the Loom v3 surface.
//!
//! Walks source line-by-line and emits a `ScannedLine` for every
//! non-blank line, classified by its leading tokens so the parser can
//! stitch them into the AST without re-scanning characters.

import { declarationKindFromKeyword } from "./ast.ts";
import { strip } from "./comments.ts";
import { Code, errorDiagnostic, type Diagnostic } from "./diagnostics.ts";
import { pos, span, type Span } from "./source.ts";
import {
  allChars,
  isAsciiAlphanumeric,
  splitInclusive,
  splitTopLevelCommas,
  stripPrefix,
  stripSuffix,
} from "./rust.ts";

export type LineKind =
  | { kind: "heading"; title: string }
  | { kind: "property"; key: string; value: string }
  | { kind: "knotMarker"; name: string; params: string[] }
  | { kind: "sceneHeading"; text: string }
  | { kind: "letBinding"; name: string; expression: string }
  | { kind: "declarationOpener"; kindWord: string; name: string; mixin: string[] }
  | { kind: "choice"; sticky: boolean; text: string }
  | { kind: "divertLine"; text: string }
  | { kind: "tunnelReturn" }
  | { kind: "speaker"; text: string }
  | { kind: "parenthetical"; text: string }
  | { kind: "fence"; tail: string; inlineClose: boolean }
  | { kind: "directive"; text: string }
  | { kind: "prose"; text: string };

/** One classified non-blank source line. */
export interface ScannedLine {
  line: number;
  indent: number;
  text: string;
  startByte: number;
  endByte: number;
  kind: LineKind;
}

export function scannedLineSpan(l: ScannedLine): Span {
  return span(
    pos(l.line, l.indent, l.startByte),
    pos(l.line, l.indent + l.text.length, l.endByte),
  );
}

interface RawScanLine {
  line: number;
  indent: number;
  text: string;
  startByte: number;
  endByte: number;
}

/** Scan `source` into classified non-blank lines plus diagnostics. */
export function scan(source: string): [ScannedLine[], Diagnostic[]] {
  const [stripped, diagnostics] = strip(source);
  source = stripped;

  // Pass 1: split into non-blank physical lines with indent + spans.
  const raws: RawScanLine[] = [];
  let byte = 0;
  let lineIdx = 0;
  for (const raw of splitInclusive(source, "\n")) {
    const lineStartByte = byte;
    const lineLen = raw.length;
    byte += lineLen;

    let strippedLine = raw.endsWith("\n") ? raw.slice(0, raw.length - 1) : raw;
    strippedLine = strippedLine.endsWith("\r")
      ? strippedLine.slice(0, strippedLine.length - 1)
      : strippedLine;

    const [indent, contentOffset] = leadingIndent(strippedLine);
    let trimmed = strippedLine.slice(contentOffset);
    trimmed = trimmed.replace(/\s+$/u, "");
    if (trimmed.length === 0) {
      lineIdx += 1;
      continue;
    }

    if (strippedLine.slice(0, contentOffset).includes("\t")) {
      diagnostics.push(
        errorDiagnostic(
          Code.L1001TabIndent,
          span(
            pos(lineIdx, 0, lineStartByte),
            pos(lineIdx, contentOffset, lineStartByte + contentOffset),
          ),
          "indentation uses tabs; Loom v3 indents with spaces only",
        ),
      );
    }

    const startByte = lineStartByte + contentOffset;
    const endByte = startByte + trimmed.length;
    raws.push({ line: lineIdx, indent, text: trimmed, startByte, endByte });
    lineIdx += 1;
  }

  // Pass 2: stitch multi-line parentheticals, then classify.
  const lines: ScannedLine[] = [];
  let i = 0;
  while (i < raws.length) {
    const opener = raws[i]!;
    let text = opener.text;
    let endByte = opener.endByte;
    let consumed = 1;
    if (opener.text.startsWith("(") && parenDepth(opener.text) > 0) {
      let depth = parenDepth(opener.text);
      let lastLine = opener.line;
      let j = i + 1;
      while (depth > 0) {
        const cont = raws[j];
        if (cont === undefined) break;
        if (cont.line !== lastLine + 1 || cont.indent < opener.indent) break;
        text += " " + cont.text;
        depth += parenDepth(cont.text);
        endByte = cont.endByte;
        lastLine = cont.line;
        j += 1;
        consumed += 1;
      }
    }
    const kind = classify(text, opener.line, opener.startByte, diagnostics);
    lines.push({
      line: opener.line,
      indent: opener.indent,
      text,
      startByte: opener.startByte,
      endByte,
      kind,
    });
    i += consumed;
  }
  return [lines, diagnostics];
}

/** Net parenthesis depth — `(` count minus `)` count. */
function parenDepth(text: string): number {
  let depth = 0;
  for (const ch of text) {
    if (ch === "(") depth += 1;
    else if (ch === ")") depth -= 1;
  }
  return depth;
}

/** Leading-whitespace columns + the offset where content begins. */
function leadingIndent(line: string): [number, number] {
  let cols = 0;
  let off = 0;
  for (const ch of line) {
    if (ch === " " || ch === "\t") {
      cols += 1;
      off += ch.length;
    } else {
      break;
    }
  }
  return [cols, off];
}

function classify(
  text: string,
  line: number,
  startByte: number,
  diagnostics: Diagnostic[],
): LineKind {
  // Header heading: `# Title` (not `##`).
  {
    const rest = stripPrefix(text, "#");
    if (rest !== null && !rest.startsWith("#")) {
      return { kind: "heading", title: rest.trim() };
    }
  }

  // Knot marker: `== name`.
  {
    const rest = stripPrefix(text, "==");
    if (rest !== null) {
      const raw = rest.trim();
      const [name, params] = splitKnotNameAndParams(raw);
      if (name.length === 0) {
        diagnostics.push(
          errorDiagnostic(
            Code.L1004UnnamedKnot,
            lineSpan(line, startByte, text),
            "`==` knot marker is missing a name",
          ),
        );
      }
      return { kind: "knotMarker", name, params };
    }
  }

  // Scene heading: `INT.`, `EXT.`, …
  if (isSceneHeading(text)) {
    return { kind: "sceneHeading", text };
  }

  // `let name = expr` reactive binding.
  {
    const rest = stripPrefix(text, "let ");
    if (rest !== null) {
      const eq = rest.indexOf("=");
      if (eq >= 0) {
        return {
          kind: "letBinding",
          name: rest.slice(0, eq).trim(),
          expression: rest.slice(eq + 1).trim(),
        };
      }
    }
  }

  // Divert / tunnel-return.
  {
    const rest = stripPrefix(text, "->");
    if (rest !== null) {
      return { kind: "divertLine", text: rest.trim() };
    }
  }
  if (text === "<-") {
    return { kind: "tunnelReturn" };
  }

  // Choice — `*` or `+` followed by required whitespace + text.
  {
    const rest = stripPrefix(text, "*");
    if (rest !== null) {
      const body = stripPrefix(rest, " ");
      if (body !== null) {
        return { kind: "choice", sticky: false, text: body.replace(/^\s+/u, "") };
      }
      if (rest.length === 0) {
        diagnostics.push(
          errorDiagnostic(
            Code.L1003EmptyChoice,
            lineSpan(line, startByte, text),
            "`*` choice has no body text",
          ),
        );
        return { kind: "choice", sticky: false, text: "" };
      }
    }
  }
  {
    const rest = stripPrefix(text, "+");
    if (rest !== null) {
      const body = stripPrefix(rest, " ");
      if (body !== null) {
        return { kind: "choice", sticky: true, text: body.replace(/^\s+/u, "") };
      }
    }
  }

  // Whole-line angle-bracket directive — `<kind: args>` or `<kind>`.
  if (text.startsWith("<") && text.endsWith(">") && text.length >= 2 && text !== "<-") {
    const inner = text.slice(1, text.length - 1);
    return { kind: "directive", text: inner };
  }

  // Triple-backtick fence.
  {
    const rest = stripPrefix(text, "```");
    if (rest !== null) {
      const inlineClose = rest.endsWith("```") && rest.length >= 3;
      const tail = inlineClose ? rest.slice(0, rest.length - 3) : rest;
      return { kind: "fence", tail, inlineClose };
    }
  }

  // Declaration opener — `KEYWORD Name [is X, Y]`.
  {
    const opener = parseDeclarationOpener(text);
    if (opener !== null) {
      if (opener.name.length === 0) {
        diagnostics.push(
          errorDiagnostic(
            Code.L1006UnnamedDeclaration,
            lineSpan(line, startByte, text),
            `\`${opener.kindWord}\` declaration is missing a name`,
          ),
        );
      }
      if (opener.unterminatedMixin) {
        diagnostics.push(
          errorDiagnostic(
            Code.L1008UnterminatedMixinClause,
            lineSpan(line, startByte, text),
            `\`is\` clause must fit on one line — a wrapped clause is dropped`,
          ),
        );
      }
      return {
        kind: "declarationOpener",
        kindWord: opener.kindWord,
        name: opener.name,
        mixin: opener.mixin,
      };
    }
  }

  // Parenthetical line — the whole line is wrapped in `(…)`.
  {
    const inner = stripParens(text);
    if (inner !== null) {
      return { kind: "parenthetical", text: inner };
    }
  }

  // Property line — `key: value`.
  {
    const prop = propertySplit(text);
    if (prop !== null) {
      return { kind: "property", key: prop[0], value: prop[1] };
    }
  }

  // Speaker cue — pure ALL CAPS.
  if (isSpeakerLine(text)) {
    return { kind: "speaker", text };
  }

  return { kind: "prose", text };
}

function lineSpan(line: number, startByte: number, text: string): Span {
  return span(pos(line, 0, startByte), pos(line, text.length, startByte + text.length));
}

function isSceneHeading(text: string): boolean {
  const upper = text.replace(/^\s+/u, "");
  return (
    upper.startsWith("INT.") ||
    upper.startsWith("EXT.") ||
    upper.startsWith("INT/EXT") ||
    upper.startsWith("INT./EXT.") ||
    upper.startsWith("I/E ")
  );
}

function stripParens(text: string): string | null {
  const a = stripPrefix(text, "(");
  if (a === null) return null;
  const inner = stripSuffix(a, ")");
  if (inner === null) return null;
  // Reject `(x)y(z)` where parens aren't balanced as the outer wrapper.
  let depth = 0;
  for (let idx = 0; idx < inner.length; idx++) {
    const ch = inner[idx];
    if (ch === "(") {
      depth += 1;
    } else if (ch === ")") {
      if (depth === 0 && idx + 1 !== inner.length) {
        return null;
      }
      depth -= 1;
    }
  }
  return inner;
}

function propertySplit(text: string): [string, string] | null {
  const colon = text.indexOf(":");
  if (colon < 0) return null;
  const key = text.slice(0, colon);
  if (key.length === 0 || !allChars(key, isPropertyKeyChar)) return null;
  // Require whitespace (or EOL) after the colon — rejects URL-shaped lines.
  const after = text.slice(colon + 1);
  if (after.length > 0 && !/^\s/u.test(after)) return null;
  return [key.trim(), after.trim()];
}

function isPropertyKeyChar(ch: string): boolean {
  return isAsciiAlphanumeric(ch) || ch === "_" || ch === "-";
}

/** Split `ask_about(topic, NPC)` into `["ask_about", ["topic", "NPC"]]`. */
function splitKnotNameAndParams(raw: string): [string, string[]] {
  raw = raw.trim();
  const open = raw.indexOf("(");
  if (open >= 0) {
    const name = raw.slice(0, open).trim();
    const tail = raw.slice(open + 1);
    const close = tail.lastIndexOf(")");
    const inner = close >= 0 ? tail.slice(0, close) : tail;
    const params = inner
      .split(",")
      .map((s) => s.trim())
      .filter((s) => s.length > 0);
    return [name, params];
  }
  return [raw, []];
}

function isSpeakerLine(text: string): boolean {
  let sawLetter = false;
  for (const ch of text) {
    if (ch >= "A" && ch <= "Z") {
      sawLetter = true;
    } else if ((ch >= "0" && ch <= "9") || ch === "_" || ch === " " || ch === "|") {
      // allowed
    } else {
      return false;
    }
  }
  return sawLetter;
}

interface DeclarationOpenerParse {
  kindWord: string;
  name: string;
  mixin: string[];
  /** True if the `is`-clause looks wrapped (trailing `,` / unbalanced `()`). */
  unterminatedMixin: boolean;
}

function parseDeclarationOpener(text: string): DeclarationOpenerParse | null {
  const m = /\s/u.exec(text);
  let kindWord: string;
  let rest: string;
  if (m) {
    kindWord = text.slice(0, m.index);
    rest = text.slice(m.index + m[0].length).trim();
  } else {
    kindWord = text;
    rest = "";
  }
  if (declarationKindFromKeyword(kindWord) === null) return null;

  const isIdx = rest.indexOf(" is ");
  let name: string;
  let mixinClause: string | null;
  if (isIdx >= 0) {
    name = rest.slice(0, isIdx).trim();
    mixinClause = rest.slice(isIdx + 4).trim();
  } else {
    name = rest;
    mixinClause = null;
  }

  // Split at top-level commas so a multi-arg trait application
  // (`CellWatch(loc: Internet, signal: lockdown)`) stays one entry — a comma
  // inside the argument list is not a separator between applications.
  const mixin =
    mixinClause !== null
      ? splitTopLevelCommas(mixinClause)
          .map((s) => s.trim())
          .filter((s) => s.length > 0)
      : [];

  // The `is`-clause is exactly one physical line (spec §2.2a). A trailing
  // top-level comma or unbalanced parens means it was wrapped — the remainder
  // would be silently dropped into the body, so flag it instead.
  let unterminatedMixin = false;
  if (mixinClause !== null) {
    const trimmed = mixinClause.trim();
    let depth = 0;
    for (const ch of trimmed) {
      if (ch === "(") depth += 1;
      else if (ch === ")") depth -= 1;
    }
    if (depth !== 0 || trimmed.endsWith(",")) unterminatedMixin = true;
  }

  return { kindWord, name, mixin, unterminatedMixin };
}
